lib/apollo-next.ts

```diff
@@ -0,0 +1,57 @@
+// "library code - would be imported from an experimental library"
+
+import { ApolloClient, useQuery as _useQuery, useMutation as _useMutation } from "@apollo/client";
+import { invariant } from "@apollo/client/utilities/globals";
+import { requestAsyncStorage } from "next/dist/client/components/request-async-storage";
+
+const apolloSymbol = Symbol.for("ApolloClient");
+
+interface ApolloSingletons {
+  client?: ApolloClient<unknown>;
+  // TODO also add suspense cache here and have that registered
+}
+
+declare module "next/dist/client/components/request-async-storage" {
+  export interface RequestStore {
+    [apolloSymbol]?: ApolloSingletons;
+  }
+}
+
+let globalMakeClient: () => ApolloClient<any> = () => {
+  invariant(false, "You need to call `registerApolloClient` before using the global Apollo Client");
+};
+
+export function registerApolloClient(makeClient: () => ApolloClient<any>) {
+  globalMakeClient = makeClient;
+}
+
+const clientApolloSingletons: ApolloSingletons = {};
+
+export const getClient = (): ApolloClient<any> => {
+  let apolloSingletons: ApolloSingletons;
+  if (typeof window !== "undefined") {
+    if (!(globalThis as any).AsyncLocalStorage) {
+      // we are in a context where `requestAsyncStorage` is not available, we have no way of doing singletons - create a new client
+      return globalMakeClient();
+    }
+    const store = requestAsyncStorage.getStore();
```

Comment:
This is the mechanism that next uses internally to make stuff like `headers` available for the whole render, so we could piggy-back in that to have a per-request `client` instance.

If this proves to be unsuccessful, going with an approach like [described in this article](https://medium.com/trabe/asynclocalstorage-for-easy-context-passing-in-node-js-e33c84679516) could also be an option (but it essentially does the same).